### DISCUSSION TITLE
Rule Trigger on "stat-" messages

### DISCUSSION THREAD

**Discussion by ajithvasudevan** [arendst/Tasmota#10467] [Q&A] [answered]
Hi all,

This is my first post here and it is about something I'd like to see in Tasmota that I couldn't find after a couple hours of searching on the internet.

I would like to have the IPAddress (as output by the "IPAddress1" command) sent to me via Telegram or a Websend.

Due to there being no "ipaddress" or similar variable I can use in a Rule, as an alternative, I would like to access the Stat- message JSON payload in Rules, just like the Tele-<SENSOR>#<value>  events.

Case in point:  If I want to get the IPAddress sent to me via Telegram or a Websend, I could just have rules defined something like -
on stat#IPAddress do tmsend %value% endon
on System#Boot do IPAddress1 endon

The second rule above would send a Stat- message with the IP Address and the first rule will then trigger and send it via Telegram/Websend.

Maybe there's another way that exists to do this. But I couldn't find it so far. If there is way, please do let me know. (Again, I re-iterate that my intention is to use Telegram or a websend rather than MQTT)

Regards,
Ajith

**Comment by barbudor** [comment, marked as answer]
Hi @ajithvasudevan 

Good logic but the `tele-` prefix exist only for tele messages.
For general response to a command you don't need to add a prefix. The whole JSON payload of the response is passed to the rule engine and the top field of the JASON are the 1st level of trigger.
As JSON response for command `IpAddress` is:
```
{ "IPAddress1": "x.y.z.t x.yz.t", "IPAddress2": ....}
```
Your trigger should be `on IPAddress1` do .... %value% endon`
Note that for a command coming for a rule to correctly trigger another rule, the command should be wrapped in `backlog`
Your full rule should be :
```
Rule1
  on system#boot do backlog ipaddress1 endon
  on IPAddress1 do .... %value% endon
```

Cheers

**Reply by sfromis** [reply]
No, due to the quirks of JSON parsing with only one key, that won't work either :-) See my response coming right after yours.

**Reply by barbudor** [reply]
But it works on my device

**Reply by sfromis** [reply]
Not on mine:
on system#boot do backlog ipaddress1 endon on IPAddress1 do var1 %value% endon"
11:12:12.325 RUL: SYSTEM#BOOT performs "backlog ipaddress1"
11:12:12.375 MQT: xxxxxxxx/stat/IPADDRESS = {"IPAddress1":"(IP unset) 192.168.xx.xx"}
11:12:12.384 RUL: IPADDRESS1 performs "var1 "
11:12:12.394 MQT: xxxxxxxx/stat/VAR = {"Var1":""}

**Reply by barbudor** [reply]
ok, you're right it doesn't work with IPAdress1, only with IPAddress

**Reply by sfromis** [reply]
I've fallen into that trap myself :-)

**Reply by barbudor** [reply]
that's weird

**Reply by sfromis** [reply]
JSON with only one key is a special case, possibly to match how it works with triggers without associated JSON published.

**Comment by sfromis** [comment]
This works:
on IPAddress1 do tmsend %value% endon 
on System#Boot do Backlog IPAddress endon

Backlog because not all commands get their JSON parsed without this, when used from a rule.
Command IPAddress because JSON parsing is more straightforward when the JSON has more han one field.
You could also have command IPAddress1 and trigger IPAddress1#data, but the latter only works if the JSON only has one key.

**Comment by ajithvasudevan** [comment]
Perfect!
@sfromis, thanks for that solution, works like a charm. Also, thank you for
the explanation.

@Barbudor, I tried what you suggested first, and, like sfromis predicted,
it didn't work. But your post gave me hope ... thanks :)

I think this kind of usage of rules (i.e., accessing JSON payload values)
is very powerful, and needs to be documented well. I can help with the
documentation, but I don't know where to start.

**Comment by meingraham** [comment]
IPADDRESS1#Data as the trigger should be the trigger

**Reply by ajithvasudevan** [reply]
Yes, that works too, if the command in the other rule was IPAddress1, instead of IPAddress. This was pointed out earlier by @sfromis

Thanks

**Comment by meingraham** [comment]
`#Data` is the "supported" trigger for _single_ level JSON _**command**_ `RESULT` output. It's [different](https://tasmota.github.io/docs/Rules/#rule-trigger) for other telemetry messages.

>Every command with a JSON payload response has an associated rule trigger.

Trigger | When it occurs
-- | --
<command>#Data | A one level JSON payload such as `{"Command":"value"}`. For example, for `{"Fanspeed":3}`, the trigger is `Fanspeed#Data`.
<command>#level1#level2#levelN | A multi-level JSON payload such as `{"TriggerLevel1":{"TriggerLevel2":{"ValueName":"value"}}}` does NOT have the `#Data` trigger. Instead, the trigger for these responses is `TriggerLevel1#TriggerLevel2#ValueName`.

**Reply by ajithvasudevan** [reply]
It is crystal clear now. I'm sorry I missed that while reading the documentation. Thank you for bringing it to my attention.
